docker/docker_image_src.go

```diff
@@ -28,17 +29,70 @@ type dockerImageSource struct {
 	cachedManifestMIMEType string // Only valid if cachedManifest != nil
 }
 
-// newImageSource creates a new ImageSource for the specified image reference.
-// The caller must call .Close() on the returned ImageSource.
-func newImageSource(sys *types.SystemContext, ref dockerReference) (*dockerImageSource, error) {
-	c, err := newDockerClientFromRef(sys, ref, false, "pull")
+// newImageSource creates a new `ImageSource` for the specified image reference
+// `ref`.
+//
+// During the creation of the instance, all provided registry mirrors within
+// `/etc/containers/registries.conf` (the default location) will be tested for
+// the availability of the image manifest. References which contain a
+// configured prefix will be automatically rewritten to the correct target
+// reference.
+//
+// If no mirror is available in the config, then the given `ref` will be used
+// to verify that the image is available at the remote location.
+//
+// A cleanup call to `.Close()` is needed if the caller is done using the returned
+// `ImageSource`.
+func newImageSource(ctx context.Context, sys *types.SystemContext, ref dockerReference) (*dockerImageSource, error) {
+	registry, err := sysregistriesv2.FindRegistry(sys, ref.ref.Name())
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrapf(err, "error loading registries configuration")
+	}
+
+	if registry == nil {
+		// No registry was found by the provided reference, so we create a
+		// fallback registry by hand to make the client creation below work as
+		// intended.
+		registry = &sysregistriesv2.Registry{
+			Endpoint: sysregistriesv2.Endpoint{
+				URL: ref.ref.String(),
+			},
+		}
 	}
-	return &dockerImageSource{
-		ref: ref,
-		c:   c,
-	}, nil
+
+	// Found the registry within the sysregistriesv2 configuration. Now we test
+	// all endpoints for the manifest availability. If an working image source
+	// was found, it will be used for all future pull actions.
+	var (
+		imageSource     *dockerImageSource
+		manifestLoadErr error
+	)
+	for _, endpoint := range append(registry.Mirrors, registry.Endpoint) {
+		newRef, err := endpoint.RewriteReference(ref.ref, registry.Prefix)
+		if err != nil {
+			return nil, errors.Wrapf(err, "error in newImageSource")
+		}
+		dockerRef := dockerReference{newRef}
+
+		client, err := newDockerClientFromRef(sys, dockerRef, false, "pull")
+		if err != nil {
+			return nil, err
+		}
+		client.endpointSkipVerify = endpoint.Insecure
+
+		testImageSource := &dockerImageSource{
+			ref: dockerRef,
+			c:   client,
+		}
+
+		manifestLoadErr = testImageSource.ensureManifestIsLoaded(ctx)
+		if manifestLoadErr == nil {
+			imageSource = testImageSource
+			break
+		}
+	}
```

Comment:
We are missing to fallback to using the initial `ref` if none of the endpoints work. This is pretty important in case the mirrors or the rewritten reference are not available for whatever reason.

Reply:
This is a pretty critical point of the mirror-semantics. I need to check if we may need to extend the Skopeo tests to catch the upper case.

Reply:
Okay, I changed the implementation to add a fallback for the original ref. I also added an additional test to the skopeo integration test PR, which now success because of this fallback:

```go
func (s *SkopeoSuite) TestSuccessCopySrcWithPrefixAndWorkingFallback(c *check.C) {
	dir := s.injectRegistriesConf(c)
	defer s.cleanupRegistriesConfInject(c, dir)

	assertSkopeoSucceeds(c, "", "copy",
		"docker://gcr.io/google-containers/busybox", "dir:"+dir)
}
```
```toml
[[registry]]
url = "gcr.io"
prefix = "gcr.io/google-containers"
```

Reply:
I need a head nod from @mtrmac on the logic as I want to make sure I did not make wrong assumptions.

@mtrmac, consider the following `registries.conf`:
```toml
[[registry]]
url = "example.com"
prefix = "docker.io"
[[registry.mirror]]
url = "some.mirror.com"
```

Assume we pull `docker.io/librarly/alpine:latest` and that the manifest is not present on `some.mirror.com` and `example.com` ... shall we fall back to `docker.io` (i.e., the initial reference) or not?

I asked @saschagrunert to add that but am now beginning to doubt that (apologies, @saschagrunert). The main thought driving my doubts is that the registry endpoint should be the last one, also due to the prefix semantics.

Reply:
LGTM